### ISSUE TITLE
Write timeouts on slow connections

### ISSUE THREAD

**Issue by GuillaumeDelente** [square/okio#294]
This issue is related to #196 

## Problem
Write timeout will still be thrown even if more than `TIMEOUT_WRITE_SIZE` bytes have been written by the delegate `sink` during the last timeout window.

## Cause
While we are splitting writes into small chunks and resetting the timeout window after calling `write` on the delegate, it might be buffering the data. As a result, when the `write` call eventually blocks, the usual write timeout will be applied even though we might be processing much more data than `TIMEOUT_WRITE_SIZE`.

## Repro steps
- Using Retrofit + OkHttp + okio, making a multipart POST (or any decent size request)
- Using a proxy to throttle the connection (128KiB/s or anything slightly bigger than the `TIMEOUT_WRITE_SIZE`) and look at the data being sent.
- Write timeout set to 1 second.

Approximately 800KiB are written to the underlying Socket before it starts blocking (12-13 write loops). At this point, we can assume that most of the 800KiB data buffered still needs to be written on the wire by the Socket. Because the upload speed is capped, the buffer will not be emptied in time in order for the timeout window to be reset, causing a timeout to be thrown.

## Potential Solution
For sockets at least, it looks like we should be able to use the send buffer size (`SO_SNDBUF`) to set `TIMEOUT_WRITE_SIZE` and potentially have every write call block (not tested yet).

**Comment by swankjesse** [comment]
I’m reluctant to add more configuration here. In addition to the local buffer in the TCP stack, it's possible that there are other buffers in routers and other middle boxes.

**Comment by GuillaumeDelente** [comment]
Yep, then maybe we shouldn't support write timeouts for sockets? This took me months to debug- we are seeing a lot of these on production (with only ~10% of which are legitimate timeouts).

**Comment by swankjesse** [comment]
Wow! How are your timeouts configured? I’d expect them to detect and fail when a server isn’t reading what a client is writing.

**Comment by GuillaumeDelente** [comment]
We have the following configuration for the OkHttp client:

```
.connectTimeout(20, TimeUnit.SECONDS)
.readTimeout(1, TimeUnit.MINUTES)
.writeTimeout(1, TimeUnit.MINUTES)
```

as you can see, it's already very generous. We're uploading files directly to S3, so I don't expect the server to be behave incorrectly.

**Comment by swankjesse** [comment]
Whoa, and 90% of the time that 20-second timeouts trigger they’re false-positives? That means that no additional bytes can be enqueued in 20 seconds. I find that highly surprising.

**Comment by GuillaumeDelente** [comment]
TL;DR
Even though this issue is a real issue, it is not what's causing my timeouts.

It's mainly the write timeout that is being thrown (which is 1 minute).
That being said, I disabled write timeout entirely and we started to get `SSLException: I/O error during system call, Connection timed out` instead, which means that something is indeed going wrong somewhere at a lower level than Okio.
To check if a timeout is legitimate or not, I was logging the result of a quick connectivity test (ping google.com/generate_204) and in 90% of the cases it succeeds- but apparently this was not the right way to do that :)
I still need to figure out what's going on with the sockets hanging, I find it hard to believe that they are all due to a network disconnect/reconnect, but that is outside of the scope of this issue.
@swankjesse thank you for your help!
